File: clone-graph.py
from __future__ import annotations
from typing import Dict, Optional

class Node:
    def __init__(self, val: int = 0, neighbors: Optional[Node] = None):
        self.val = val
        self.neighbors = neighbors or []

class Solution:
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        if node is None:
            return None
        
        def clone(node: Optional[Node], hashmap: Dict[int, Node]) -> Optional[Node]:
            if node is None:
                return None

            copy = Node(node.val)
            hashmap[node.val] = copy
            
            for neighbor in node.neighbors:
                if neighbor.val not in hashmap.keys():
                    clone(neighbor, hashmap)
                copy.neighbors.append(hashmap[neighbor.val])

            return copy

        return clone(node, {})

File: clone-graph.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*<();
    }

    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<int, Node*> hashmap;
        return clone(node, hashmap);
    }

    Node* clone(Node* node, unordered_map<int, Node*> hashmap) {
        if (node == NULL) return NULL;

        Node* copy = new Node(node->val);
        hashmap[node->val] = copy;

        for (Node* neighbor: node->neighbors) {
            if (hashmap.count(neighbor->val) == 0) {
                clone(neighbor, hashmap);
            }
            copy->neighbors.push_back(hashmap[neighbor->val]);
        }

        return copy;
    }
};
